### PR TITLE
ignore measure when warmup

### PR DESCRIPTION
**Pull request by Connor1996**
When I bench, I find that max latency is always high at the very beginning and can't observe such high latency from TiKV grpc duration metrics. Then I print some log find the latency is usually high for the first time to send request. So this PR make it not to record the first request latency.

### CODE REVIEW

**Comment by siddontang** [comment]
it is bad to let the Measurement know the WarmUp, it is the responsibility for the test client.

I think we need:

1. Add a configuration to decide the warmup time, not only just once.
2. After WarmUp, the measurement and counter can start working.

File: pkg/workload/core.go [hunk 1]
```diff
@@ -456,17 +452,10 @@ func (c *core) doTransactionReadModifyWrite(ctx context.Context, db ycsb.DB, sta
...
-		return err
-	}
-
-	if err := db.Update(ctx, c.table, keyName, values); err != nil {
-		return err
-	}
-
-	if c.dataIntegrity {
-		c.verifyRow(state, keyName, readValues)
+	if t, ok := db.(*client.DbWrapper); ok {
```
Comment (siddontang) [inline pkg/workload/core.go]: very wired that the core knows client module.

File: pkg/client/client.go [hunk 2]
```diff
@@ -151,27 +154,51 @@ func (w *worker) run(ctx context.Context) {
...
+	wg.Add(threadCount)
+	measureCtx, measureCancel := context.WithCancel(ctx)
+	go func() {
+		// load stage no need to warm up
+		if c.p.GetBool(prop.DoTransactions, true) {
+			dur := c.p.GetInt64("warmuptime", 0)
+			time.Sleep(time.Duration(dur) * time.Second)
+		}
+		// finish warming up
+		c.db.WarmUp = false
```
Comment (siddontang) [inline pkg/client/client.go]: data race for WarmUp var

Comment (Connor1996) [reply pkg/client/client.go]: I have noticed that, but it doesn't matter a lot. `WarmUp` is only set once here and for read threads it is acceptable to get a slightly old value.

File: pkg/workload/core.go (hunk 1 above)
Comment (Connor1996) [reply pkg/workload/core.go]: I agree. But if not, we need to add ReadModifyWrite to DB interface and every DB need to implement that.

Comment (siddontang) [reply pkg/workload/core.go]: please check how BatchDB does

File: pkg/client/client.go [hunk 3]
```diff
@@ -151,27 +154,51 @@ func (w *worker) run(ctx context.Context) {
...
+	var wg sync.WaitGroup
 	threadCount := c.p.GetInt(prop.ThreadCount, 100)
 
-	c.wg.Add(threadCount)
+	wg.Add(threadCount)
+	measureCtx, measureCancel := context.WithCancel(ctx)
+	go func() {
+		// load stage no need to warm up
+		if c.p.GetBool(prop.DoTransactions, true) {
+			dur := c.p.GetInt64("warmuptime", 0)
```
Comment (siddontang) [inline pkg/client/client.go]: if we close outside, here, the goroutine must wait WarmUp time then exit.

File: pkg/client/client.go (hunk 2 above)
Comment (siddontang) [reply pkg/client/client.go]: use atomic to avoid data race.

File: pkg/ycsb/db.go [hunk 4]
```diff
@@ -70,6 +70,15 @@ type DB interface {
 	Delete(ctx context.Context, table string, key string) error
 }
 
+type AtomicDB interface {
```
Comment (siddontang) [inline pkg/ycsb/db.go]: this is not an Atomic operation, just read, modify and write.

File: pkg/ycsb/db.go [hunk 5]
```diff
@@ -70,6 +70,16 @@ type DB interface {
 	Delete(ctx context.Context, table string, key string) error
 }
 
+// WrappedDB gives some combined operations based on basic DB Operations.
+type WrappedDB interface {
```
Comment (siddontang) [inline pkg/ycsb/db.go]: bad name

Comment (Connor1996) [reply pkg/ycsb/db.go]: any suggested?

File: pkg/workload/core.go [hunk 6]
```diff
@@ -456,17 +451,10 @@ func (c *core) doTransactionReadModifyWrite(ctx context.Context, db ycsb.DB, sta
...
-	}
-
-	if err := db.Update(ctx, c.table, keyName, values); err != nil {
-		return err
-	}
-
-	if c.dataIntegrity {
-		c.verifyRow(state, keyName, readValues)
+	if t, ok := db.(ycsb.WrappedDB); ok {
+		t.ReadModifyWrite(ctx, c.table, keyName, fields, values)
```
Comment (zhangjinpeng1987) [inline pkg/workload/core.go]: Need to verifyRow here?

File: pkg/client/dbwrapper.go [hunk 7]
```diff
@@ -24,10 +25,15 @@ import (
 
 // DbWrapper stores the pointer to a implementation of ycsb.DB.
 type DbWrapper struct {
-	DB ycsb.DB
+	DB     ycsb.DB
+	WarmUp int32 // use as bool, 1 for true, 0 for false
```
Comment (zhangjinpeng1987) [inline pkg/client/dbwrapper.go]: 1 means in WarmUp progressing.

File: pkg/ycsb/db.go (hunk 5 above)
Comment (siddontang) [reply pkg/ycsb/db.go]: I am confused why do you want to provide this Interface?

File: pkg/client/client.go [hunk 8]
```diff
@@ -164,12 +166,41 @@ func NewClient(p *properties.Properties, workload ycsb.Workload, db ycsb.DB) *Cl
...
+		if c.p.GetBool(prop.DoTransactions, true) {
+			dur := c.p.GetInt64(prop.WarmUpTime, 0)
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(time.Duration(dur) * time.Second):
+			}
+		}
+		// finish warming up
+		measurement.FinishWarmUp()
```
Comment (siddontang) [inline pkg/client/client.go]: if we don't enable Warmup, but the default WarmUp is true, so when executing this, the following thread executions may run for a time, but they previous executed must be measured.

File: pkg/measurement/measurement.go [hunk 9]
```diff
@@ -82,7 +83,11 @@ func InitMeasure(p *properties.Properties) {
 	globalMeasure = new(measurement)
 	globalMeasure.p = p
 	globalMeasure.opMeasurement = make(map[string]ycsb.Measurement, 16)
-	warmUp = 1
+	if p.GetInt64(prop.WarmUpTime, 0) <= 0 {
```
Comment (siddontang) [inline pkg/measurement/measurement.go]: please provide an `EnableWarmUp(b bool)` function instead of using global var directly.

**Pull request merged**